File: StriverA2Z/JavaScript&TypeScript/src/Arrays/Medium/kadaneAlgorithmMaximumSumSubArray.ts
function maxSubArray(arr: number[]): number {
	let sum = 0;
	let max = Number.MIN_SAFE_INTEGER;
	let start = 0;
	let ansEnd = -1, ansStart = -1;
	for (let i = 0; i < arr.length; i++) {
		if (sum == 0) start = i;
		sum += arr[i];
		if (sum > max) {
			max = sum;
			ansStart = start;
			ansEnd = i;
		}
		if (sum < 0) {
			sum = 0;
		}
	}

	for (let i = ansStart; i <= ansEnd; i++) {
		console.log(arr[i], " ");
	}
	return max;
}

console.log(maxSubArray([-2, 1, -3, 4, -1, 2, 1, -5, 4])); // 6

File: StriverA2Z/C++/3_ArraysQuestion/3.2_Medium/kadane_algorithm_maximum_sum_subarray.cpp
#include<iostream>

using namespace std;

void kadaneAlgorithm(int arr[], int n) {
    int maximum = INT_MIN;
    int sum = 0;
    int start = 0;
    int ansStart = -1, ansEnd = -1;
    for (int i = 0; i < n; i++) {
        if (sum == 0) start = i;
        sum += arr[i];
        if (sum > maximum) {
            maximum = sum;
            ansStart = start;
            ansEnd = i;
        }
        if (sum < 0) {
            sum = 0;
        }
    }

    for (int i = ansStart; i <= ansEnd; i++) {
        cout << arr[i] << " ";
    }
    cout << endl << "Maximum sum of subarray is " << maximum;
}

int main() {
    int arr[] = {-2, -3, 4, -1, -2, 1, 5, -3};
    kadaneAlgorithm(arr, sizeof(arr) / sizeof(arr[0]));
    return 0;
}